src/nft.sol

```diff
@@ -32,166 +32,163 @@ contract IdentityFactoryLike {
-      * @dev Parses a uint and returns the hex string
-      * @param payload uint
-      * @return string the corresponding hex string
-      */
-     function uintToHexStr(uint payload) internal pure returns (string memory) {
-       if (payload == 0)
-         return "0";
-       // calculate string length
-       uint i = payload;
-       uint length;
-
-       while (i != 0) {
-         length++;
-         i = i >> 4;
-       }
-       // parse byte by byte and construct the string
-       i = payload;
-       uint mask = 15;
-       bytes memory result = new bytes(length);
-       uint k = length - 1;
-
-       while (i != 0) {
-         uint curr = (i & mask);
-         result[k--] = curr > 9 ? byte(55 + uint8(curr)) : byte(48 + uint8(curr));
-         i = i >> 4;
-       }
-
-       return string(result);
-     }
-
-    function equalBytes(bytes memory a, bytes memory b) internal pure returns (bool) {
-        if (a.length == b.length) {
-            for (uint i = 0; i < a.length; i++) {
-                if (a[i] != b[i]) {
-                    return false;
-                }
-            }
-            return true;
+    /**
+     * Computes the asset hash from the given params and verifies if its present in the AssetManager
+     * @param to address deposit address of the token
+     * @param properties bytes[] properties of the each proof
+     * @param values bytes[] value associated with each property
+     * @param salts bytes[] salt associated with each property
+     */
+    function _verifyAsset(address to, bytes[] memory properties, bytes[] memory values, bytes32[] memory salts) internal view  {
+        // construct assetHash from the props, values, salts
+        // append to address
+        bytes memory data = abi.encodePacked(to);
+
+        // append hashes
+        for (uint i=0; i< properties.length; i++) {
+            data = abi.encodePacked(data, keccak256(abi.encodePacked(properties[i], values[i], salts[i])));
         }
-        return false;
+
+        assetManager.verify(keccak256(data));
+    }
+
+    /**
+     * @dev Checks that provided document is signed by the given identity. Validates and checks if the public key used is a not revoked and has the purpose SIGNING_KEY.
+     * Does not check if the signature root is part of the document root.
+     * @param identity address identity which is a collaborator of the document
+     * @param dataRoot bytes32 hash of all data fields of the document which are signed
+     * @param signature bytes contains signature + transition flag
+     */
+    function _signed_document(address identity, bytes32 dataRoot, bytes memory signature) internal view {
+        // check that the identity being used has been created by the Centrifuge Identity Factory contract
+        require(identityFactory.createdIdentity(identity), "nft/identity-not-registered");
+        // extract flag from signature
+        (bytes memory sig, byte flag) = _recoverSignatureAndFlag(signature);
```

Comment:
why are you not passing the `flag` as an extra parameter?

What's the benefit in splitting the flag from the signature on-chain?

Reply:
Because it is encoded in the signature leaf of the document already.